### PR TITLE
Improving the timeseries suite list interface

### PR DESCRIPTION
**Pull request by ledm**
Initial PR for this patch.

It should do the first half of the issue #39:
- list of switches to turn on and off individual keys

It's a set of individual files which contain a list of flags.

### CODE REVIEW

**Comment by ledm** [comment]
Think this one is ready for review.

**Comment by ledm** [comment]
Just need check that the commented lists match the new files. fun times.

File: bgcval2/UKESMpython.py [hunk 1]
```diff
@@ -71,8 +71,8 @@ def folder(name):
     """ This snippet takes a string, makes the folder and the string.
 	    It also accepts lists of strings.
 	"""
-    if type(name) == type(['a', 'b', 'c']):
-        name = join(name, '/')
+    if isinstance(name, list):
+        name = os.path.join(*name) # splat!
```
Comment (valeriupredoi) [inline bgcval2/UKESMpython.py:75]:
```suggestion
        name = os.path.join(*name)
```

File: bgcval2/analysis_compare.py [hunk 2]
```diff
@@ -183,163 +181,14 @@ def timeseries_compare(jobs,
...
-
-#               	analysisKeys.append('quickSST')    		# Area Weighted Mean Surface Temperature
-#       	  	analysisKeys.append('TotalOMZVolume')           # Total Oxygen Minimum zone Volume
-#       	 	analysisKeys.append('OMZThickness')             # Oxygen Minimum Zone Thickness
-#        	analysisKeys.append('OMZMeanDepth')             # Oxygen Minimum Zone mean depth
-#		analysisKeys.append('O2')                       # WOA Oxygen
-#       	if bio ==False:return
-#       	if physics == True:return
+    if isinstance(suites, str):
+        suites = [suites, ]
```
Comment (valeriupredoi) [inline bgcval2/analysis_compare.py:188]: add an `else:
                  ValueError(f"Suites need to be a string, got {suites}")
                   sys.exit(1)`

File: bgcval2/analysis_compare.py [hunk 3]
```diff
@@ -183,163 +181,14 @@ def timeseries_compare(jobs,
...
+
+    analysisKeys = build_list_of_suite_keys(suites, debug = True)
```
Comment (valeriupredoi) [inline bgcval2/analysis_compare.py:190]:
```suggestion
    analysisKeys = build_list_of_suite_keys(suites, debug=True)
```

File: bgcval2/analysis_compare.py [hunk 4]
```diff
@@ -183,163 +181,14 @@ def timeseries_compare(jobs,
...
+    print('analysisKeys', analysisKeys)
```
Comment (valeriupredoi) [inline bgcval2/analysis_compare.py:191]:
```suggestion
    print(f'Using analysis keys {str(analysisKeys)}')
```

File: bgcval2/analysis_compare.py [hunk 5]
```diff
@@ -4260,7 +4001,7 @@ def load_yml_and_run(compare_yml, config_user):
 
     # Master suite leys:
     if not master_suites:
-        master_suites=['physics', 'bio'] # Defaults
+        master_suites=['physics', 'bgc'] # Defaults
```
Comment (valeriupredoi) [inline bgcval2/analysis_compare.py:4004]:
```suggestion
        master_suites=['physics', 'bgc']  # Defaults
```

File: bgcval2/analysis_timeseries.py [hunk 6]
```diff
@@ -252,9 +88,61 @@ def listModelDataFiles(jobID, filekey, datafolder, annual):
     return model_files
 
 
+def build_list_of_suite_keys(suites, debug=True):
+    """
+    Generate a list of keys from a list of suites.
+
+    """
+    print('build_list_of_suite_keys')
```
Comment (valeriupredoi) [inline bgcval2/analysis_timeseries.py:96]:
```suggestion
    print('analysis_timeseries: Calling build_list_of_suite_keys to build list of keys')
```

File: bgcval2/analysis_timeseries.py [hunk 7]
```diff
@@ -252,9 +88,61 @@ def listModelDataFiles(jobID, filekey, datafolder, annual):
...
+    paths_dir = os.path.dirname(os.path.realpath(__file__))
+    key_lists_dir = os.path.join(os.path.dirname(paths_dir), 'key_lists')
+
+    print('key_lists_dir', key_lists_dir)
```
Comment (valeriupredoi) [inline bgcval2/analysis_timeseries.py:100]:
```suggestion
    print(f'analysis_timeseries: Directory where keys are stored: {key_lists_dir}')
```

File: bgcval2/analysis_timeseries.py [hunk 8]
```diff
@@ -252,9 +88,61 @@ def listModelDataFiles(jobID, filekey, datafolder, annual):
...
+    analysis_keys = {}
+    for suite in suites:
+        print(suite)
+        # look for a list in keys_list directory:
+        suite_yml = os.path.join(key_lists_dir, ''.join([suite.lower(),'.yml']))
+        if debug:
+            print('build_list_of_suite_keys:\tlooking for suite yaml:', suite_yml)
+
+        if not os.path.exists(suite_yml):
+            print('build_list_of_suite_keys:\tERROR: suite yaml:', suite_yml, 'does not exist')
```
Comment (valeriupredoi) [inline bgcval2/analysis_timeseries.py:110]:
```suggestion
            print(f'analysis_timeseries: build_list_of_suite_keys:\tERROR: suite yaml: {suite_yml} file does not exist')
```

Comment (valeriupredoi) [reply bgcval2/analysis_timeseries.py:110]: ideally you should use Python builtin errors and raise them, OK for now

File: bgcval2/analysis_timeseries.py [hunk 9]
```diff
@@ -252,9 +88,61 @@ def listModelDataFiles(jobID, filekey, datafolder, annual):
...
+        if not dictionary or not isinstance(dictionary, dict):
+            print(f"Configuration file {suite_yml} "
+                  "is either empty or corrupt, please check its contents")
+            sys.exit(1)
+
+        keys_dict = dictionary.get('keys', {})
+
+        for key, keybool in keys_dict.items():
+            if debug and key in analysis_keys:
+                print('build_list_of_suite_keys:\tKey exists in multiple suites:', key)
```
Comment (valeriupredoi) [inline bgcval2/analysis_timeseries.py:126]:
```suggestion
                print(f'build_list_of_suite_keys:\tKey {key} exists in multiple suites:')
```

File: bgcval2/analysis_timeseries.py [hunk 10]
```diff
@@ -252,9 +88,61 @@ def listModelDataFiles(jobID, filekey, datafolder, annual):
...
+
+
+            if key in analysis_keys and keybool != analysis_keys[key]:
+                print('build_list_of_suite_keys:\tERROR: conflick in input yamls:', key, keybool, '!=', analysis_keys[key])
```
Comment (valeriupredoi) [inline bgcval2/analysis_timeseries.py:130]:
```suggestion
                print(f'build_list_of_suite_keys:\tERROR: conflict in input yamls: {key}, {keybool} != {analysis_keys[key]}')
```

File: bgcval2/analysis_timeseries.py [hunk 11]
```diff
@@ -288,9 +176,9 @@ def analysis_timeseries(
     print('-----------------------')
     print('Starting analysis_timeseries')
     print('jobID:', jobID)
-    print('analysisSuite:',analysisSuite)
+    print('suites:', suites)
     print('regions:', regions)
-    print('clean:',  clean, 'annual:',annual, 'strictFileCheck:', strictFileCheck)
+    print('clean:',  clean, 'annual:', annual, 'strictFileCheck:', strictFileCheck)
```
Comment (valeriupredoi) [inline bgcval2/analysis_timeseries.py:181]:
```suggestion
    print(f'clean: {clean},  annual: {annual}, strictFileCheck: {strictFileCheck}')
```

Comment (valeriupredoi) [reply bgcval2/analysis_timeseries.py:181]: get used with the f-string formatting, much cleaner, and Pythonic :grin:

File: bgcval2/analysis_timeseries.py [hunk 12]
```diff
@@ -309,183 +197,12 @@ def analysis_timeseries(
...
-#            'u-ak900',
-#            'u-ar538',
-#            'u-an869',
-#            'u-ar977',
-#    ]:
-#        try:
-#            analysisKeys.remove('FreshwaterFlux')
-#        except:
-#            pass
+    analysisKeys = build_list_of_suite_keys(suites, debug = True)
```
Comment (valeriupredoi) [inline bgcval2/analysis_timeseries.py:204]:
```suggestion
    analysisKeys = build_list_of_suite_keys(suites, debug=True)
```

File: key_lists/bgc.yml [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    # Nutrients
+    N: True  # WOA Nitrate
+    Si: True  # WOA Siliate
+    Iron: True  # Iron
+
+    # Production, Export, Chlorophyll and DMS
+    CHL: True  # Total Chlorophyll
+    # CHD: True  # Diatom Chlorophyll
+    # CHN: True  # Non-diatom Chlorophyll
+    #	DiaFrac: True                  # Diatom Fraction
```
Comment (valeriupredoi) [inline key_lists/bgc.yml:11]:
```suggestion
    # DiaFrac: True  # Diatom Fraction
```

File: key_lists/bgc.yml [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+    IntPP_OSU: True  # OSU Integrated primpary production
+    PP_OSU: True  # OSU Integrated primpary production
+    # DMS_ARAN: True
+    # DMS: True                      
+    # DTC: True                      # Detrital carbon
+    #	LocalExportRatio: True         # Export ratio (no data)
```
Comment (valeriupredoi) [inline key_lists/bgc.yml:18]:
```suggestion
    # LocalExportRatio: True         # Export ratio (no data)
```

File: key_lists/bgc.yml [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    #	GlobalExportRatio: True        # Export ratio (no data)
```
Comment (valeriupredoi) [inline key_lists/bgc.yml:19]:
```suggestion
    # GlobalExportRatio: True        # Export ratio (no data)
```

File: key_lists/bgc.yml [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    Alk: True  # Glodap Alkalinity
+    DIC: True  # Globap tCO2
+    AirSeaFluxCO2: True  # Air Sea Flux
+    TotalAirSeaFluxCO2: True  # Total global air sea flux
+    pH: True
+
+    # Oxygen:
+    O2: True  # WOA Oxygen
+    TotalOMZVolume: True  # Total Oxygen Minimum zone Volume
+    #	OMZThickness: True             # Oxygen Minimum Zone Thickness
```
Comment (valeriupredoi) [inline key_lists/bgc.yml:31]: careful how you format the commented out lines - if someone who doesn't know yaml uncomments these they will immediately run into yaml loading issues coxz the indentation is wrong - I always leave either a single space between # and text or none at all

**Review by valeriupredoi — APPROVED** [review]
cool! I would add a short description to each of those yaml files at the top via a couple commented out lines; see my minor corrections, merge at will, bud! :beer:

File: bgcval2/UKESMpython.py (hunk 1 above)
Comment (ledm) [reply bgcval2/UKESMpython.py:75]: I've never used splat before, so I added this comment to remind me what that operator does.

File: bgcval2/analysis_compare.py (hunk 2 above)
Comment (ledm) [reply bgcval2/analysis_compare.py:188]: Actually, suites need to be a list! I'll amend it here.

**Comment by ledm** [comment]
Approved changes, added comments to README. Happy to merge and move onto next part.

**Pull request merged**